calyx-py/calyx/builder.py

```diff
@@ -0,0 +1,341 @@
+        elif key.endswith('_'):
+            self.ctx_asgn(key[:-1], value)
+        else:
+            self.ctx_asgn(key, value)
+
+
+class GroupBuilder:
+    """Wraps a group for easy addition of assignment statements.
+
+    The basic mechanism here is the `asgn` method, which builds a single
+    assignment statement and adds it to the underlying group. The `done`
+    property also provides access to the group's done hole.
+
+    There is also a fancy, magical way to add assignments using Python
+    assignment syntax based on Python's context managers. Use `with` on
+    a `GroupBuilder` to enter a context where assignments can
+    *implicitly* get added to this group.
+    """
+
+    def __init__(self, group: ast.Group, comp: ComponentBuilder):
+        self.group = group
+        self.comp = comp
+
+    def asgn(self, lhs: ExprBuilder,
+             rhs: ExprBuilder | CondExprBuilder | int,
+             cond=None):
+        """Add a connection to the group.
+
+        If the assigned value is an int, try to infer a width for it and
+        promote it to a constant expression. If it's a `CondExprBuilder`
+        (which you create as (`cond @ value`), use the condition
+        contained therein for the assignment.
+        """
+
+        if isinstance(rhs, CondExprBuilder):
+            assert cond is None
+            cond = rhs.cond
+            rhs = rhs.value
+
+        if isinstance(rhs, int):
+            width = infer_width(lhs)
+            assert width, f'could not infer width for literal {rhs}'
+            rhs = const(width, rhs)
+
+        wire = ast.Connect(
+            ExprBuilder.unwrap(lhs),
+            ExprBuilder.unwrap(rhs),
+            ExprBuilder.unwrap(cond),
+        )
+        self.group.connections.append(wire)
+
+    @property
+    def done(self):
+        """The `done` hole for the group."""
+        return ExprBuilder(
+            ast.HolePort(ast.CompVar(self.group.id.name), "done")
+        )
+
+    @done.setter
+    def done(self, expr):
+        """Build an assignment to `done` in the group."""
+        self.asgn(self.done, expr)
+
+    def __enter__(self):
+        TLS.groups.append(self)
+        return self
+
+    def __exit__(self, exc, value, tb):
+        TLS.groups.pop()
+
+
+def const(width: int, value: int):
+    """Build a sized integer constant expression.
+
+    This is available as a shorthand in cases where automatic width
+    inference fails. Otherwise, you can just use plain Python integer
+    values.
+    """
+    return ast.ConstantPort(width, value)
+
+
+def infer_width(expr):
+    """Try to guess the width of a port expression.
+
+    Return an int, or None if we don't have a guess.
+    """
+    assert TLS.groups, "int width inference only works inside `with group:`"
+    group_builder = TLS.groups[-1]
+
+    # Deal with `done` holes.
+    expr = ExprBuilder.unwrap(expr)
+    if isinstance(expr, ast.HolePort):
+        assert expr.name == 'done', f"unknown hole {expr.name}"
```

Comment:
Just a side thought; have you thought about having a list of "legal" names for ports as well? This would avoid cases like `reg.read_data = 1;`. For the Stdlib, it could be hard-coded in the Python AST. For new primitives, it could be some opt-in thing.

Reply:
Yeah, that's an interesting idea. Having some of these "signatures" available in an easy-to-evolve way would be rad. I'll add it to the to-do list.